ts/metrics/metric_cache_abstract.py

```diff
@@ -0,0 +1,462 @@
+            name=name,
+            value=value,
+            req_id=req_id,
+            unit=unit,
+            metric_type=metric_type,
+            dimensions=dimensions,
+        )
+
+    def add_percent(
+        self,
+        name: str,
+        value: int or float,
+        idx=None,
+        dimensions: list = None,
+        metric_type: MetricTypes = MetricTypes.gauge,
+    ):
+        """
+        Add a percentage based metric
+            Default metric type is gauge
+
+        Parameters
+        ----------
+        name : str
+            metric name
+        value: int, float
+            value of metric
+        idx: int
+            request_id index in batch
+        dimensions: list
+            list of dimensions for the metric
+        metric_type: MetricTypes
+           type for defining different operations, defaulted to gauge metric type for Percent metrics
+        """
+        unit = "percent"
+        req_id = self._get_req(idx)
+        self._add_or_update(
+            name=name,
+            value=value,
+            req_id=req_id,
+            unit=unit,
+            metric_type=metric_type,
+            dimensions=dimensions,
+        )
+
+    def add_error(
+        self,
+        name: str,
+        value: int or float,
+        dimensions: list = None,
+        metric_type: MetricTypes = MetricTypes.counter,
+    ):
+        """
+        Add an Error Metric
+            Default metric type is counter
+
+        Parameters
+        ----------
+        name : str
+            metric name
+        value: int, float
+            value of metric, in this case a str
+        dimensions: list
+            list of dimensions for the metric
+        metric_type: MetricTypes
+           type for defining different operations, defaulted to counter metric type for Error metrics
+        """
+        unit = ""
+        # noinspection PyTypeChecker
+        self._add_or_update(
+            name=name,
+            value=value,
+            req_id=None,
+            unit=unit,
+            metric_type=metric_type,
+            dimensions=dimensions,
+        )
+
+    def get_metric(
+        self, metric_type: MetricTypes, metric_name: str, dimensions: list or str
+    ) -> Metric:
+        """
+        Get a Metric from cache.
+            Ask user for required requirements to form metric key to retrieve Metric.
+
+        Parameters
+        ----------
+        metric_type: MetricTypes
+            Type of metric: use MetricTypes enum to specify
+
+        metric_name: str
+            Name of metric
+
+        dimensions: list or str
+            list of dimension keys which should be strings
+
+        """
+        metric_key = f"[{metric_type.value}]-[{metric_name}]-[{dimensions}]"
+
+        metric_obj = self.cache.get(metric_key)
+        if metric_obj:
+            logging.debug(f"Successfully received metric {metric_key}")
+            return metric_obj
+        else:
+            raise merrors.MetricsCacheKeyError(
```

Comment:
Do we need to have the custom wrappers on errors? If we throw an error then we should also see a callstack which would show this line of code

Reply:
Think it's more of a nice-to-have, specifying/narrowing down the area of what the error is + giving a bit more context